provider/plugin/plugin.go

```diff
@@ -0,0 +1,350 @@
+	Current  string `json:"current"`
+}
+
+type PropertyValuesEqualResponse struct {
+	Equals bool `json:"equals"`
+}
+
+func init() {
+	prometheus.MustRegister(recordsErrorsGauge)
+	prometheus.MustRegister(applyChangesErrorsGauge)
+	prometheus.MustRegister(propertyValuesEqualErrorsGauge)
+	prometheus.MustRegister(adjustEndpointsErrorsGauge)
+}
+
+func NewPluginProvider(u string) (*PluginProvider, error) {
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// negotiate API information
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
+
+	client := &http.Client{}
+	var resp *http.Response
+	err = backoff.Retry(func() error {
+		resp, err = client.Do(req)
+		if err != nil {
+			log.Debugf("Failed to connect to plugin api: %v", err)
+			return err
+		}
+		// we currently only use 200 as success, but considering okay all 2XX for future usage
+		if resp.StatusCode >= 300 && resp.StatusCode < 500 {
+			return backoff.Permanent(fmt.Errorf("status code < 500"))
+		}
+		return nil
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to plugin api: %v", err)
+	}
+
+	vary := resp.Header.Get(varyHeader)
+	contentType := resp.Header.Get(contentTypeHeader)
+
+	if vary != contentTypeHeader {
+		return nil, fmt.Errorf("wrong vary value returned from server: %s", vary)
+	}
+
+	if contentType != mediaTypeFormatAndVersion {
+		return nil, fmt.Errorf("wrong content type returned from server: %s", contentType)
+	}
+
+	return &PluginProvider{
+		client:          client,
+		remoteServerURL: parsedURL,
+	}, nil
+}
+
+// Records will make a GET call to remoteServerURL/records and return the results
+func (p PluginProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to join path: %s", err.Error())
+		return nil, err
+	}
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to create request: %s", err.Error())
+		return nil, err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
+	resp, err := p.client.Do(req)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to perform request: %s", err.Error())
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to get records with code %d", resp.StatusCode)
+		return nil, fmt.Errorf("failed to get records with code %d", resp.StatusCode)
+	}
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to read response body: %s", err.Error())
+		return nil, err
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+	err = json.Unmarshal(b, &endpoints)
```

Comment:
The response can get quite big. It should stream the unmarshal.

Reply:
Same as https://github.com/kubernetes-sigs/external-dns/pull/3063#discussion_r1219100964 .

Reply:
`err = json.NewDecoder(resp.Body).Decode(&endpoints)`